### PR TITLE
Metadata tag

### PR DESCRIPTION
**Pull request by kalaspuffar**
Issue https://github.com/mtmse/dotify.formatter.impl/issues/25

Description:
Enable the addition of a metadata tag in OBFL to have it available in the PEF output.

Changes:
* Added insertMetadata as a function to add a new segment.

### CODE REVIEW

**Comment by kalaspuffar** [comment]
Hi @bertfrees and @PaulRambags

Please review this PR. It is connected to the issue above, and the actual implementation is in the formatter.

Best regards
Daniel

**Comment by bertfrees** [comment]
Hi @kalaspuffar. A few remarks:

- I think the name "metadata" isn't so great. You expect "metadata" to be metadata about the whole document, and that is indeed what it is already used for. Not in the [`Formatter`](https://mtmse.github.io/dotify.api/5.0.3/javadoc/org/daisy/dotify/api/formatter/Formatter.html) interface, but in the [`ObflParser`](https://mtmse.github.io/dotify.api/5.0.3/javadoc/org/daisy/dotify/api/obfl/ObflParser.html) interface. Especially in OBFL it will be confusing because you will have `meta` and `metadata` in the same file and they will be rather different things. I can see where you are coming from with the name "metadata" though. It kind of _is_ metadata, but not in the usual sense (attached to a document or element) but attached to a single very specific position in the document. I have to admit I don't immediately have a suggestion for a good name. It's a pretty generic feature.
- I would expect a change also in the [writer](https://mtmse.github.io/dotify.api/5.0.3/javadoc/org/daisy/dotify/api/writer/package-summary.html) API?
- This is kind of related to OBFL, although I would try to keep the two things as separate as possible: the object that you pass is a `Map`. This is a pretty safe choice because it's a generic data structure and can easily be converted to and from XML. On the other hand the data that you pass through the formatter using this feature will never be truely "generic" in the sense that you will always know where the data comes from, so you could express that in the API by making the argument an `Object` (and let the same object come out on the other end). Either you are reading/writing the data directly from/to your Java application in which case you know the type of the object at all times. Or you are writing to a PEF, in which case you know that if metadata is present it must either come from a OBFL, or from an application that knows it is writing to a PEF, so the object can be expected to be a map.

   You could also go the other direction, and change the map into a simple `String`. This will be sufficient information. The reasoning here is that whoever needs to interpret the output (through PEF or Java API or other way) is also the one who constructed the input (through OBFL or Java API or other way) so he can use the string as a key to lookup other information if needed. Such lookups will probably be needed anyway even when more data can be passed through the formatter using the map.

   All three approaches work. The slight concern I have with the map is that it is so clearly geared towards OBFL input and PEF output.

   If we stick with map and don't care about the connection with XML I think it might be a good idea to change it to `Map<QName,String>` because that models better a list of XML attributes.

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

Well, we could look into changing the metadata tag to BlockMeta perhaps? That would make it more specific and distinguish it from the head meta tag.

I read through all your remarks but I might have missed some subtleties, it's always easier with examples but your comment mostly boils down to that we need to be more specific with the API so people know what's expected to be sent. So a Map<QName, String> would make that more specific?

Best regards 
Daniel

**Comment by bertfrees** [comment]
> your comment mostly boils down to that we need to be more specific with the API so people know what's expected to be sent

No not quite :-)

What I'm basically trying to say that if it were up to me I would probably go for `Object`. If find the map too geared towards OBFL/PEF. We should always keep in mind that this API tries to be an abstraction of that.

On the other hand I know that not everyone is comfortable using Object in their (Java) API. So that is why I also suggested `String`.

My last sentence is just to make sure you are aware that _if_ you are indeed trying to model a list of XML attributes, `Map<QName,String>` is a better choice than `Map<String,String>`. But note that this ties the API even more to XML, so it is not my preferred solution.

**Comment by kalaspuffar** [comment]
Hi @bertfrees 

So you want it to change to Map<Object,Object> ?
If we go less restrive we could model more things but then we need to handle more cases that might be bad. For instance what if you send in the class Price ... how should the API handle that?

Best regards
Daniel

**Comment by bertfrees** [comment]
Just `Object.` No `Map`. The API does not need to be aware of the type. All you have to say is that the objects that are passed in should come out unchanged.

As far as the implementation goes: the formatter doesn't need to be aware of the type either. It's only the media writers that will actually need to do something with the objects. The PEF media writer could handle the case where the objects are not maps (by ignoring them or throwing an error). But as I explained this is really not expected to occur and should be considered an exception because the user of the formatter API should normally be aware that it is writing to a PEF.

Note that also with a `Map<String,String>` you'd have to do some validation because not any string represents a valid attribute name. Using `QName` would solve that though.

Sure, I know, the less typed you go, the more "manual" checking you have to do, and the more you rely on Javadoc to document the API. That is of course the downside. But it's a trade-off. I guess the proper "Java"-way to do it would be to parameterize the Formatter class.

**Comment by kalaspuffar** [comment]
Hi @bertfrees and @PaulRambags

Please review this PR again as I want to build a new API version so we can get passing tests in the formatter going forward. Of course, this might still be the wrong API, but I think we have discussed it a lot :)

Best regards
Daniel

**Review by bertfrees — APPROVED** [review]
Approved. But if you think it might not be the final one, why don't you just use a snapshot?

**Comment by bertfrees** [comment]
Hmm, maybe I was too fast. I don't know if this addition to the API makes much sense without also updating the [writer](https://mtmse.github.io/dotify.api/5.0.3/javadoc/org/daisy/dotify/api/writer/package-summary.html) API and explaining how the objects are passes right through without modification.

**Comment by kalaspuffar** [comment]
Well, there were no changes to the writer API but I could add the information about the data transfer. Any suggestions where I should add that information? @bertfrees

**Comment by bertfrees** [comment]
I mean without an update to the writer API the objects can't come out. So at the moment you can't say much useful about it in the API.

I think what you can do now is document in PEFMediaWriter that you are interpreting the objects as attributes and you are inserting these in the PEF. It currently only works if the provided `Row` are `RowImpl` though. An update of the writer API is really appropriate.

**Comment by kalaspuffar** [comment]
Ahh, @bertfrees.

I see your point. I've added an interface class to enable Writer implementors to use this external reference object in their implementations.

**Review by bertfrees — CHANGES_REQUESTED** [review]
OK that's a start. As I explained [elsewhere](https://github.com/mtmse/dotify.formatter.impl/issues/25#issuecomment-624001225), I would ideally like the API to already take into account other use cases and not restrict the info to be attached to a row only, even if that is how it will be used initially. It's better that you don't add methods that you might have to remove later. On the other hand, to find a good API in this case is not so trivial and probably would need some proper thought.

Regarding the Javadoc: it's important that we keep making the distinction between the formatter and writer APIs on the one side, and OBFL and PEF on the other side. In order to use the formatter you don't need an OBFL, and you can have other writer implementations that don't create a PEF. It's of course fine, and even good, to mention them in the docs, but as examples.

**Comment by bertfrees** [comment]
> to find a good API in this case is not so trivial 

Hmm maybe it's not so hard after all... What if `getExternalReference`(s) would return a set (or list) of references, each with a certain position (character index)? Or we could keep the `getExternalReference` method (being a reference that it is attached to the row as a whole) as it is and we keep the option to add another method later to retrieve such "inter-character" specific references?

**Comment by kalaspuffar** [comment]
Hi @bertfrees and @PaulRambags

I'm still waiting for review and approval for this PR.

Best regards
Daniel

File: src/org/daisy/dotify/api/writer/Row.java
Comment (PaulRambags) [inline src/org/daisy/dotify/api/writer/Row.java]:
```
/**
 * Get the external reference element.
 *
 * <p>The returned object should be unchanged.</p>
 *
 * <p>A use-case...       </p>
```

File: src/org/daisy/dotify/api/writer/Row.java
Comment (PaulRambags) [inline src/org/daisy/dotify/api/writer/Row.java]:
```
* @return The external reference object, not modified
```

**Review by PaulRambags — CHANGES_REQUESTED** [review]

**Review by PaulRambags — APPROVED** [review]

**Review by bertfrees — APPROVED** [review]

**Pull request merged**
